garrysmod/lua/includes/extensions/string.lua

```diff
@@ -1,92 +1,76 @@
+
+	return val
+end
+
+local function string_ToTable( str )
 	local tbl = {}
 
-	for i = 1, string.len( str ) do
-		tbl[i] = string.sub( str, i, i )
+	for i = 1, #str do
+		tbl[i] = string_sub( str, i, i )
 	end
 
 	return tbl
 end
 
---[[---------------------------------------------------------
-	Name: string.JavascriptSafe( string )
-	Desc: Takes a string and escapes it for insertion in to a JavaScript string
------------------------------------------------------------]]
+string.ToTable = string_ToTable
+
 local javascript_escape_replacements = {
 	["\\"] = "\\\\",
-	["\0"] = "\\x00" ,
-	["\b"] = "\\b" ,
-	["\t"] = "\\t" ,
-	["\n"] = "\\n" ,
-	["\v"] = "\\v" ,
-	["\f"] = "\\f" ,
-	["\r"] = "\\r" ,
+	["\0"] = "\\x00",
+	["\b"] = "\\b",
+	["\t"] = "\\t",
+	["\n"] = "\\n",
+	["\v"] = "\\v",
+	["\f"] = "\\f",
+	["\r"] = "\\r",
 	["\""] = "\\\"",
 	["\'"] = "\\\'"
 }
 
 function string.JavascriptSafe( str )
-
-	str = str:gsub( ".", javascript_escape_replacements )
-
+	str = string_gsub( str, ".", javascript_escape_replacements )
 	-- U+2028 and U+2029 are treated as line separators in JavaScript, handle separately as they aren't single-byte
-	str = str:gsub( "\226\128\168", "\\\226\128\168" )
-	str = str:gsub( "\226\128\169", "\\\226\128\169" )
-
-	return str
-
-end
-
---[[---------------------------------------------------------
-	Name: string.PatternSafe( string )
-	Desc: Takes a string and escapes it for insertion in to a Lua pattern
------------------------------------------------------------]]
-local pattern_escape_replacements = {
-	["("] = "%(",
-	[")"] = "%)",
-	["."] = "%.",
-	["%"] = "%%",
-	["+"] = "%+",
-	["-"] = "%-",
-	["*"] = "%*",
-	["?"] = "%?",
-	["["] = "%[",
-	["]"] = "%]",
-	["^"] = "%^",
-	["$"] = "%$",
-	["\0"] = "%z"
-}
+	str = string_gsub( str, "\226\128\168", "\\\226\128\168" )
 
-function string.PatternSafe( str )
-	return ( str:gsub( ".", pattern_escape_replacements ) )
+	return string_gsub( str, "\226\128\169", "\\\226\128\169" )
 end
 
---[[---------------------------------------------------------
-	Name: explode(seperator ,string)
-	Desc: Takes a string and turns it into a table
-	Usage: string.explode( " ", "Seperate this string")
------------------------------------------------------------]]
-local totable = string.ToTable
-local string_sub = string.sub
-local string_find = string.find
-local string_len = string.len
-function string.Explode(separator, str, withpattern)
-	if ( separator == "" ) then return totable( str ) end
-	if ( withpattern == nil ) then withpattern = false end
+local function string_Explode( sep, str, withpattern )
+	if ( sep == "" ) then return string_ToTable( str ) end
+	withpattern = !withpattern
 
 	local ret = {}
 	local current_pos = 1
 
-	for i = 1, string_len( str ) do
-		local start_pos, end_pos = string_find( str, separator, current_pos, !withpattern )
-		if ( !start_pos ) then break end
+	for i = 1, #str do
+		local start_pos, end_pos = string_find( str, sep, current_pos, withpattern )

```

Comment:
We can also localize `string_find` inside, but that's not a big deal.

Reply:
Will LuaJIT not optimise out the const local upval/treat it as a local?

Reply:
Hm, apparently over-localizing introduces additional equality check, probably because upvalues has const attribute internally, while locals don't.

Over-localization is a trick I use for non-JIT code